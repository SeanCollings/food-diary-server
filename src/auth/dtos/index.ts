export * from '@/auth/dtos/create-user.dto';
export * from '@/auth/dtos/login-user.dto';
export * from '@/auth/dtos/reset-password.dto';
