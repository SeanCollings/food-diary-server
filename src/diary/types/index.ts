import { MAX_CALENDAR_ENTRIES_RANGE } from '@/lib/constants/validation/validation.constants';
import { Type } from 'class-transformer';
import { IsDateString, IsNotEmpty, IsNumber, Max } from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export class GetDiaryEntriesQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class GetCalendarEntriesQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;

  @IsNotEmpty()
  @IsNumber()
  @Max(MAX_CALENDAR_ENTRIES_RANGE)
  @Type(() => Number)
  months: string;
}

export type MealColumnKeys =
  | 'mealBreakfast'
  | 'mealSnack1'
  | 'mealLunch'
  | 'mealSnack2'
  | 'mealDinner';

export type HasMealColumnKeys =
  | 'hasMealBreakfast'
  | 'hasMealSnack1'
  | 'hasMealLunch'
  | 'hasMealSnack2'
  | 'hasMealDinner';
