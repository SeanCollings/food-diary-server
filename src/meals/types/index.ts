import { IsDateString, IsNotEmpty } from 'class-validator';
import { Request } from 'express';

// https://stackoverflow.com/questions/50160662/convert-nested-array-type

export class CreateMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class UpdateMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export class DeleteMealEntryQuery {
  @IsNotEmpty()
  @IsDateString()
  date: string;
}

export enum MealTypesEnum {
  BREAKFAST = 'breakfast',
  SNACK_1 = 'snack_1',
  LUNCH = 'lunch',
  SNACK_2 = 'snack_2',
  DINNER = 'dinner',
}
export type MealTypes = `${MealTypesEnum}`;

export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export interface MealContent {
  id: string;
  food: string;
  serving?: string;
  quantity?: string;
  description?: string;
  emoji?: { name: string; nativeSkin: string } | null;
}

export type SetFieldOld<V, O extends object> = {
  [K in keyof O]: O[K] extends object ? SetFieldOld<V, O[K]> : V;
};

type Primitive = string | number | boolean | null | undefined;

type SetField<V, O> = O extends Primitive
  ? V
  : O extends (infer U)[]
  ? SetField<V, U>
  : { [K in keyof O]: SetField<V, O[K]> };

type RequireOne<T> = T & { [P in keyof T]: Required<Pick<T, P>> }[keyof T];
