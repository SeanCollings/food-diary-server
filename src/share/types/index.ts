import { MAX_SUMMARY_MONTH_RANGE } from '@/lib/constants/validation/validation.constants';
import { IsDateAfter } from '@/lib/validation/validators/is-date-after';
import { IsWithinDateRange } from '@/lib/validation/validators/is-within-date-range';
import {
  IsDateString,
  IsNotEmpty,
  IsString,
  IsUUID,
  MaxLength,
  MinLength,
} from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export class GetSharedSummaryQuery {
  @IsNotEmpty()
  @MinLength(36)
  @MaxLength(36)
  @IsString()
  @IsUUID()
  link: string;

  @IsNotEmpty()
  @IsDateString()
  dateFrom: string;

  @IsNotEmpty()
  @IsDateString()
  @IsDateAfter('dateFrom', { message: 'dateFrom must come before dateTo' })
  @IsWithinDateRange('dateFrom', {
    message: `dateFrom and dateTo max-range is ${MAX_SUMMARY_MONTH_RANGE} months`,
  })
  dateTo: string;
}
