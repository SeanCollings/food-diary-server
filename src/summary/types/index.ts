import { MAX_SUMMARY_MONTH_RANGE } from '@/lib/constants/validation/validation.constants';
import { IsDateAfter } from '@/lib/validation/validators/is-date-after';
import { IsWithinDateRange } from '@/lib/validation/validators/is-within-date-range';
import { IsDateString, IsNotEmpty } from 'class-validator';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export class GetUserSummaryQuery {
  @IsDateString()
  @IsNotEmpty()
  dateFrom: string;

  @IsDateString()
  @IsNotEmpty()
  @IsDateAfter('dateFrom', { message: 'dateFrom must come before dateTo' })
  @IsWithinDateRange('dateFrom', {
    message: `dateFrom and dateTo max-range is ${MAX_SUMMARY_MONTH_RANGE} months`,
  })
  dateTo: string;
}
