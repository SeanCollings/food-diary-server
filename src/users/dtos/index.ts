export * from '@/users/dtos/update-user.dto';
export * from '@/users/dtos/update-preferences.dto';
export * from '@/users/dtos/user.dto';
