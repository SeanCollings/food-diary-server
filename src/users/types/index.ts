import { User } from '@prisma/client';

export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export type PreferenceColumnNames = keyof Pick<
  User,
  | 'userPreferenceShowDayStreak'
  | 'userPreferenceShowWeeklyExcercise'
  | 'userPreferenceShowWeeklyWater'
>;

export type UserWithShareLink =
  | (User & {
      shareLink: {
        link: string;
        isShared: boolean;
      } | null;
    })
  | null;
