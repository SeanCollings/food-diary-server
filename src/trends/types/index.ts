import { IsIn, IsNotEmpty, IsString } from 'class-validator';
import { WellnessTypesEnum } from '@/wellness/types';

export interface RequestWithUser extends Request {
  user: { userId: number; email: string };
}

export type TrendType = 'week' | 'month';

export class GetTrendsQuery {
  @IsString()
  @IsNotEmpty()
  @IsIn(['week', 'month'])
  type: TrendType;
}

export type BeverageTypes = Exclude<
  WellnessTypesEnum,
  WellnessTypesEnum.EXCERCISE
>;
