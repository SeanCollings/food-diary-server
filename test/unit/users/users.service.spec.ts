import { PrismaService } from '@/prisma.service';
import { UpdatePreferencesDTO, UpdateUserDTO } from '@/users/dtos';
import { UserWithShareLink } from '@/users/types';
import { UsersService } from '@/users/users.service';
import { Test, TestingModule } from '@nestjs/testing';
import { DiaryDay, PrismaClient } from '@prisma/client';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';

describe('UsersService', () => {
  const mockUserId = 1234;
  const mockDateToday = '2023-04-28';
  const mockDateYesterday = '2023-04-27';
  let service: UsersService;
  let prisma: DeepMockProxy<PrismaClient>;

  const mockDiaryDays: Partial<DiaryDay>[] = [
    {
      id: '1',
      userId: mockUserId,
      date: mockDateToday,
      hasMealBreakfast: true,
      hasMealSnack1: true,
      hasMealDinner: true,
      wellnessWater: 2,
      wellnessAlcohol: 1,
      wellnessExcercise: '01:15',
      wellnessExcerciseDetails: 'mock exercise details',
    },
    {
      id: '2',
      userId: mockUserId,
      date: mockDateYesterday,
      hasMealBreakfast: true,
      hasMealLunch: true,
      wellnessWater: 4,
      wellnessTeaCoffee: 2,
    },
  ];

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [UsersService, PrismaService],
    })
      .overrideProvider(PrismaService)
      .useValue(mockDeep<PrismaClient>())
      .compile();

    service = module.get(UsersService);
    prisma = module.get(PrismaService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('findOne', () => {
    it('should find a single user by email', async () => {
      prisma.user.findUnique.mockResolvedValue({
        email: 'test@email.com',
      } as any);

      const result = await service.findOne('test@email.com');
      expect(result).toEqual({
        email: 'test@email.com',
      });
    });
  });

  describe('findOneSelect', () => {
    it('should get a user and return selected fields', async () => {
      prisma.user.findUnique.mockResolvedValue({
        name: 'test',
        email: 'test@email.com',
      } as any);

      const result = await service.findOneSelect(1234, {
        name: true,
        email: true,
      });

      expect(prisma.user.findUnique).toHaveBeenCalledWith(
        expect.objectContaining({
          select: {
            name: true,
            email: true,
          },
        }),
      );
      expect(result).toEqual({
        name: 'test',
        email: 'test@email.com',
      });
    });

    it('should cater for no select given', async () => {
      prisma.user.findUnique.mockResolvedValue({} as any);

      await service.findOneSelect(1234);
      expect(prisma.user.findUnique).toHaveBeenCalledWith(
        expect.objectContaining({
          select: {},
        }),
      );
    });
  });

  describe('getUserProfile', () => {
    const userWithShareLink: Partial<UserWithShareLink> = {
      name: 'Mock Name',
      email: 'test@email.com',
      statLastActivity: new Date(mockDateToday),
      statDayStreak: 4,
      userPreferenceShowDayStreak: true,
      shareLink: {
        isShared: true,
        link: 'mock_sharelink',
      },
    };

    it('should get user profile', async () => {
      prisma.user.findUnique.mockResolvedValue(userWithShareLink as any);
      prisma.diaryDay.findMany.mockResolvedValue(mockDiaryDays as any);

      const result = await service.getUserProfile(mockUserId);

      expect(result).toMatchSnapshot();
      expect(prisma.user.findUnique.mock.calls[0][0]).toMatchSnapshot();
      expect(prisma.diaryDay.findMany.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('updateUser', () => {
    const mockUserUpdate: UpdateUserDTO = { name: 'Mock Name' };

    it('should update user', async () => {
      prisma.user.update.mockResolvedValue({} as any);
      await service.updateUser(1234, mockUserUpdate);
      expect(prisma.user.update.mock.calls[0][0]).toMatchSnapshot();
    });
  });

  describe('updateUserPreferences', () => {
    const mockUserPreferences: UpdatePreferencesDTO = {
      showDayStreak: true,
      showWeeklyExcercise: true,
      showWeeklyWater: true,
    };

    it('should update users preferences', async () => {
      prisma.user.update.mockResolvedValue({} as any);

      await service.updateUserPreferences(mockUserId, mockUserPreferences);
      expect(prisma.user.update.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should update users preferences without preferences', async () => {
      await service.updateUserPreferences(mockUserId, {});
      expect(prisma.user.update).not.toHaveBeenCalled();
    });
  });

  describe('updateUserStreak', () => {
    it('should update a users streak if last activity date was yesterday', async () => {
      prisma.user.findUnique.mockResolvedValue({
        statLastActivity: new Date(mockDateYesterday),
        statDayStreak: 4,
      } as any);
      prisma.user.update.mockResolvedValue({} as any);

      await service.updateUserStreak(mockUserId);

      expect(prisma.user.findUnique).toBeCalledTimes(1);
      expect(prisma.user.update.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should not update a users streak if last activity date was today', async () => {
      prisma.user.findUnique.mockResolvedValue({
        statLastActivity: new Date(mockDateToday),
        statDayStreak: 4,
      } as any);
      prisma.user.update.mockResolvedValue({} as any);

      await service.updateUserStreak(mockUserId);

      expect(prisma.user.findUnique).toBeCalledTimes(1);
      expect(prisma.user.update).not.toBeCalled();
    });

    it('should cater for an undefined statDayStreak value', async () => {
      prisma.user.findUnique.mockResolvedValue({
        statLastActivity: new Date(mockDateYesterday),
      } as any);
      prisma.user.update.mockResolvedValue({} as any);

      await service.updateUserStreak(mockUserId);

      expect(prisma.user.findUnique).toBeCalledTimes(1);
      expect(prisma.user.update.mock.calls[0][0]).toMatchSnapshot();
    });

    it('should cater for null found user', async () => {
      prisma.user.findUnique.mockResolvedValue(null);

      await service.updateUserStreak(mockUserId);

      expect(prisma.user.findUnique).toBeCalledTimes(1);
      expect(prisma.user.update).not.toBeCalled();
    });
  });
});
