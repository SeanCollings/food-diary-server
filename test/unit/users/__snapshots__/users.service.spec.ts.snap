// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersService getUserProfile should get user profile 1`] = `
{
  "avatar": undefined,
  "email": "test@email.com",
  "name": "Mock Name",
  "preferences": {
    "isProfileShared": true,
    "showDayStreak": true,
    "showWeeklyExcercise": undefined,
    "showWeeklyWater": undefined,
  },
  "shareLink": "mock_sharelink",
  "stats": {
    "dayStreak": 4,
    "weeklyExercise": 75,
    "weeklyWater": 6,
  },
}
`;

exports[`UsersService getUserProfile should get user profile 2`] = `
{
  "include": {
    "shareLink": {
      "select": {
        "isShared": true,
        "link": true,
      },
    },
  },
  "where": {
    "id": 1234,
  },
}
`;

exports[`UsersService getUserProfile should get user profile 3`] = `
{
  "where": {
    "AND": {
      "date": {
        "in": [
          "2023-04-21T22:00:00.000Z",
          "2023-04-22T22:00:00.000Z",
          "2023-04-23T22:00:00.000Z",
          "2023-04-24T22:00:00.000Z",
          "2023-04-25T22:00:00.000Z",
          "2023-04-26T22:00:00.000Z",
          "2023-04-27T22:00:00.000Z",
        ],
      },
    },
    "userId": 1234,
  },
}
`;

exports[`UsersService updateUser should update user 1`] = `
{
  "data": {
    "name": "Mock Name",
  },
  "where": {
    "id": 1234,
  },
}
`;

exports[`UsersService updateUserPreferences should update users preferences 1`] = `
{
  "data": {
    "userPreferenceShowDayStreak": true,
    "userPreferenceShowWeeklyExcercise": true,
    "userPreferenceShowWeeklyWater": true,
  },
  "where": {
    "id": 1234,
  },
}
`;

exports[`UsersService updateUserStreak should cater for an undefined statDayStreak value 1`] = `
{
  "data": {
    "statDayStreak": 1,
    "statLastActivity": 2023-04-27T22:00:00.000Z,
  },
  "where": {
    "id": 1234,
  },
}
`;

exports[`UsersService updateUserStreak should update a users streak if last activity date was yesterday 1`] = `
{
  "data": {
    "statDayStreak": 5,
    "statLastActivity": 2023-04-27T22:00:00.000Z,
  },
  "where": {
    "id": 1234,
  },
}
`;
